Documentation for spacy-curated-transformers (#12677)
explosion/spaCy: website/docs/api/curated-transformer.mdx

> ---

Reviewer: We typically avoid hyphens in the name of the page, so I'd suggest renaming to `/api/curatedtransformer.mdx`.

Also, we need to add this page to the index of the API pages or it won't show up in the menu on the left.

Reviewer 2: Renamed and added.

title: CuratedTransformer
teaser: Pipeline component for multi-task learning with transformer models
tag: class
source: github.com/explosion/spacy-transformers/blob/master/spacy_curated_transformers/pipeline_component.py

> version: 3

Reviewer: Can't we be more specific here?

Reviewer 2: Changed to `3.7`.

api_base_class: /api/pipe
api_string_name: transformer
---

> #### Installation
>
> ```bash
> $ pip install -U spacy-curated-transformers
> ```

<Infobox title="Important note" variant="warning">

This component is available via the extension package
[`spacy-curated-transformers`](https://github.com/explosion/spacy-curated-transformers). It
exposes the component via entry points, so if you have the package installed,
using `factory = "curated_transformer"` in your
[training config](/usage/training#config) or `nlp.add_pipe("curated_transformer")` will
work out-of-the-box.

</Infobox>

> This Python package provides a curated set of transformer models for spaCy. It is focused on deep integration into spaCy and will support deployment-focused features such as distillation and quantization. Curated transformers currently supports the following model types:

Reviewer 2: ```suggestion
This Python package provides a curated set of transformer models for spaCy. It is focused on deep integration into spaCy and will support deployment-focused features such as distillation and quantization in the future. spaCy curated transformers currently supports the following model types:
```

spaCy's [Tok2Vec](/api/tok2vec), and the
[Tok2VecListener](/api/architectures/#Tok2VecListener) sublayer. The component
assigns the output of the transformer to the `Doc`'s extension attributes. To
access the values, you can use the custom
[`Doc._.trf_data`](#assigned-attributes) attribute.

For more details, see the [usage documentation](/usage/embeddings-transformers).

## Assigned Attributes {id="assigned-attributes"}

The component sets the following
[custom extension attribute](/usage/processing-pipeline#custom-components-attributes):

| Location         | Value                                                                      |
| ---------------- | -------------------------------------------------------------------------- |
| `Doc._.trf_data` | Curated transformer outputs for the `Doc` object. ~~DocTransformerOutput~~ |

## Config and implementation {id="config"}

The default config is defined by the pipeline component factory and describes
how the component should be configured. You can override its settings via the
`config` argument on [`nlp.add_pipe`](/api/language#add_pipe) or in your
[`config.cfg` for training](/usage/training#config). See the

> [model architectures](/api/architectures#transformers) documentation for details
> on the transformer architectures and their arguments and hyperparameters.

Reviewer: ```suggestion
[model architectures](/api/architectures#curated-trf) documentation for details
on the curated transformer architectures and their arguments and hyperparameters.
```
?

Note that the default config does not include the mandatory `vocab_size`
hyperparameter as this value can differ between different models. So, you will
need to explicitly specify this before adding the pipe (as shown in the example
below).

> #### Example
>
> ```python
> from spacy_curated_transformers.pipeline.transformer import DEFAULT_CONFIG
>
> config = DEFAULT_CONFIG.copy()
> config["transformer"]["model"]["vocab_size"] = 250002
> nlp.add_pipe("curated_transformer", config=config["transformer"])
> ```

| Setting             | Description                                                                                                                                                                                                                                        |
| ------------------- | -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |

> | `model`             | The Thinc [`Model`](https://thinc.ai/docs/api-model) wrapping the transformer. Defaults to [CuratedTransformerModel](/api/architectures#CuratedTransformerModel). ~~Model[List[Doc], FullCuratedTransformerBatch]~~                                |

Reviewer: I'm not seeing where this `/api/architectures#CuratedTransformerModel` link would go to? Maybe it's not actually documented yet?

Reviewer 2: Fixed.

| `frozen`            | If `True`, the model's weights are frozen and no backpropagation is performed. ~~bool~~                                                                                                                                                            |
| `all_layer_outputs` | If `True`, the model returns the outputs of all the layers. Otherwise, only the output of the last layer is returned. This must be set to `True` if any of the pipe's downstream listeners require the outputs of all transformer layers. ~~bool~~ |

```python
https://github.com/explosion/spacy-curated-transformers/blob/main/spacy_curated_transformers/pipeline/transformer.py
```

## CuratedTransformer.\_\_init\_\_ {id="init",tag="method"}

> #### Example
>
> ```python
> # Construction via add_pipe with default model
> trf = nlp.add_pipe("curated_transformer")
>
> # Construction via add_pipe with custom config
> config = {
>     "model": {
>         "@architectures": "spacy-curated-transformers.XlmrTransformer.v1",
>         "vocab_size": 250002,
>         "num_hidden_layers": 12,

> >         "hidden_width": 768

Reviewer: ```suggestion
>         "hidden_width": 768,
```

| `frozen`            | If `True`, the model's weights are frozen and no backpropagation is performed. ~~bool~~                                                                                                                                                            |
| `all_layer_outputs` | If `True`, the model returns the outputs of all the layers. Otherwise, only the output of the last layer is returned. This must be set to `True` if any of the pipe's downstream listeners require the outputs of all transformer layers. ~~bool~~ |

## CuratedTransformer.\_\_call\_\_ {id="call",tag="method"}

Apply the pipe to one document. The document is modified in place, and returned.
This usually happens under the hood when the `nlp` object is called on a text
and all pipeline components are applied to the `Doc` in order. Both
[`__call__`](/api/transformer#call) and [`pipe`](/api/transformer#pipe) delegate
to the [`predict`](/api/transformer#predict) and

> [`set_annotations`](/api/transformer#set_annotations) methods.

Reviewer: Shouldn't all these links be updated to `/api/curatedtransformer`?

Reviewer 2: Fixed.

> #### Example
>
> ```python
> doc = nlp("This is a sentence.")
> trf = nlp.add_pipe("curated_transformer")
> # This usually happens under the hood

> > processed = transformer(doc)

Reviewer: ```suggestion
> processed = trf(doc)
```

> ```

| Name        | Description                      |
| ----------- | -------------------------------- |
| `doc`       | The document to process. ~~Doc~~ |
| **RETURNS** | The processed document. ~~Doc~~  |

## CuratedTransformer.pipe {id="pipe",tag="method"}

Apply the pipe to a stream of documents. This usually happens under the hood
when the `nlp` object is called on a text and all pipeline components are
applied to the `Doc` in order. Both [`__call__`](/api/transformer#call) and
[`pipe`](/api/transformer#pipe) delegate to the
[`predict`](/api/transformer#predict) and

> [`set_annotations`](/api/transformer#set_annotations) methods.

Reviewer: Same comment about the `api` links, best to double-check throughout.

Reviewer 2: Fixed.

`[components.transformer.model.with_spans]` block of the config to customize the
sequences processed by the transformer.

| Name        | Description                                                   |
| ----------- | ------------------------------------------------------------- |
| `docs`      | A batch of `Doc` objects. ~~Iterable[Doc]~~                   |
| **RETURNS** | The spans to process by the transformer. ~~List[List[Span]]~~ |

### WithStridedSpans.v1 {id="strided_spans",tag="registered function"}

> #### Example config
>
> ```ini
> [transformer.model.with_spans]
> @architectures = "spacy-curated-transformers.WithStridedSpans.v1"
> stride = 96
> window = 128
> ```

Create a span getter for strided spans. If you set the `window` and `stride` to
the same value, the spans will cover each token once. Setting `stride` lower
than `window` will allow for an overlap, so that some tokens are counted twice.
This can be desirable, because it allows all tokens to have both a left and
right context.

| Name     | Description              |
| -------- | ------------------------ |
| `window` | The window size. ~~int~~ |
| `stride` | The stride size. ~~int~~ |

## Tokenizer loaders

> Placeholder text for tokenizers

Reviewer: TODO

Reviewer 2: Fixed.
